### DISCUSSION TITLE
Flutter VLC Player

### DISCUSSION THREAD

**Discussion by mitchross** [alexmercerind/dart_vlc#233] [Ideas]
I worked heavily on https://github.com/solid-software/flutter_vlc_player

Theres a concept of https://docs.flutter.dev/development/platform-integration/platform-channels 

Which the flutter VLC player has https://github.com/solid-software/flutter_vlc_player/tree/master/flutter_vlc_player_platform_interface 

Im curious if @alexmercerind  has any suggestions or ideas how we can bend dart_vlc together with platform interfaces in flutter_vlc_player to allow for desktop support in flutter VLC player and vice versa allow android/ios on dart_vlc.

Still noodling here....

**Comment by alexmercerind** [comment]
Original comment by @alexmercerind at https://github.com/alexmercerind/dart_vlc/issues/157#issuecomment-1076561569
<hr></hr>

@ mitchross

> I worked heavily on https://github.com/solid-software/flutter_vlc_player

Hello @mitchross, thanks for joining the conversation. [`flutter_vlc_player`](https://github.com/solid-software/flutter_vlc_player) looks nice. I recall @ mfkl from videoLAN linked that after I posted [`dart_vlc`](https://github.com/alexmercerind/dart_vlc) on Twitter. I honestly had no idea about existing libvlc project for Flutter, either way I wanted audio playback for desktop back then. Following @ mfkl's reply, I joined the Discord for libvlc & I'm pretty sure I've seen you & you've seen me there.

My work has always been specific for Windows/Linux (C++) & since there were no video playback available libraries (still not available either for Linux or Windows), I decided to work on this. From my observations, libvlc 3.0.0's video frame callbacks aren't really performant. There are some people complaining about CPU usage on Windows, I wonder how well it'd be running on Android.
Following the performance issue #184, I started working on [`flutter_native_view`](https://github.com/alexmercerind/flutter_native_view) (currently just supporting Windows) because Flutter team couldn't make platform views work on Windows.

Current implementation of [`dart_vlc`](https://github.com/alexmercerind/dart_vlc) uses `dart:ffi` & `package:ffi` for communicating with libvlc & uses platform channel interface _*just*_ for texture registration (which I will possibly remove if I start using [`flutter_native_view`](https://github.com/alexmercerind/flutter_native_view)). And I fear mixing the two separate implementation (one based on platform interface i.e. `flutter_vlc_player` & another on ffi i.e. [`dart_vlc`](https://github.com/alexmercerind/dart_vlc).

I'm not really able to spare much time for this project these days [see commit history] & I don't know if I will be able to contribute back to [`flutter_vlc_player`](https://github.com/solid-software/flutter_vlc_player) as of now. On the other hand, I think there are many good alternatives available to libvlc (in terms of _features_, _bundle-size_ or even _performance_) such as official `winrt::` media playback APIs on Windows, ExoPlayer on Android, GStreamer on Linux etc. etc. This plugin ([`dart_vlc`](https://github.com/alexmercerind/dart_vlc)) is now just "restricted" to use libvlc because of it's name. 

No fear to say, I wonder if people opened/upvoted this issue only because they didn't wanna write separate code for Android/iOS (using `video_player` or `better_player` which also allow caching, better performance, smaller bundle sizes, provide more features & use ExoPlayer). To me, using `libvlc` on Android is bit redundant (atleast just for playing media).
There are some really really fine OSS apps on _just_ GitHub using [`dart_vlc`](https://github.com/alexmercerind/dart_vlc). From my observations, they're using it just because there are no alternatives to [`dart_vlc`](https://github.com/alexmercerind/dart_vlc). libvlc isn't really business friendly [LGPL/GPL] or good to bundle (#220). I wonder most projects will readily switch to something else once things are bit sorted out for desktop too.

Lastly, I really really appreciate you taking the initiative for the collaboration. I'm currently not really ready to drop my copyleft LGPL on this project. Still, feel free to depend upon [`dart_vlc`](https://github.com/alexmercerind/dart_vlc) as a `pub.dev` dependency in your project or projects.

@ zdabing 
> can you use libmpv make dart_mpv. In my opinion 

I already have a private `libmpv` library used in [Harmonoid](https://github.com/harmonoid). An early-POC & public version is available here under permissive MIT license: [libmpv.dart](https://github.com/alexmercerind/libmpv.dart). The version used in [Harmonoid](https://github.com/harmonoid) is not available under open-source license as of now (capable of YouTube support, tag parsing & video playback).

TL;DR

I'm not going to implement Android support in [`dart_vlc`](https://github.com/alexmercerind/dart_vlc) (atleast anytime soon), because I have no revenue out of this, boss shouting at me to do this or any paycheck for this & I'm tired while having other plugins to maintain too.
However, I'll be happy to review/improve any pull requests that one makes here for Android support.

Thanks.
I think I've been little too honest with the reply, regardless have a great day.
